### PR TITLE
Messaging Queue Share

### PR DESCRIPTION
**Pull request by problemv**
For #836 

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

https://github.com/openstack/zaqar/blob/28f1f0f9370c395023bc5ae01693e5d952b3e1c1/zaqar/transport/wsgi/v2_0/urls.py#L43

### CODE REVIEW

**Comment by problemv** [comment]
@jtopjian Rebased and ready to go.

File: openstack/messaging/v2/queues/requests.go [hunk 1]
```diff
@@ -196,3 +196,38 @@ func GetStats(client *gophercloud.ServiceClient, queueName string) (r StatResult
...
+type ShareOpts struct {
+	Paths   []string `json:"paths,omitempty"`
+	Methods []string `json:"methods,omitempty"`
+	Expires string   `json:"expires,omitempty"`
+}
+
+// ShareOptsBuilder allows extensions to add additional attributes to the
+// Share request.
+type ShareOptsBuilder interface {
+	ToShareQueueMap() (map[string]interface{}, error)
```
Comment (jtopjian) [inline openstack/messaging/v2/queues/requests.go]: `ToQueuesShareMap`

The format is:

To-Resources-Action-Result

File: openstack/messaging/v2/queues/requests.go [hunk 2]
```diff
@@ -196,3 +196,38 @@ func GetStats(client *gophercloud.ServiceClient, queueName string) (r StatResult
 		OkCodes: []int{200}})
 	return
 }
+
+// ShareOpts specifies share creation parameters.
+type ShareOpts struct {
+	Paths   []string `json:"paths,omitempty"`
```
Comment (jtopjian) [inline openstack/messaging/v2/queues/requests.go]: It looks like there are only a limited number of paths:

https://github.com/openstack/zaqar/blob/28f1f0f9370c395023bc5ae01693e5d952b3e1c1/zaqar/transport/wsgi/v2_0/urls.py#L31

We _could_ make a custom type for these, but that feels kind of weird to me for some reason. If you feel a custom type would be beneficial, by all means go for it, but I'm also good with leaving this as `[]string`.

Comment (problemv) [reply openstack/messaging/v2/queues/requests.go]: I am going to leave this as string. The path would have the dynamic queue name in it. I will however update the unit tests to reflect this being the case.

```/v2/queues/test/subscriptions```

**Comment by problemv** [comment]
@jtopjian Changes implemented. I decided to set up a custom type for both Path and Methods.

File: acceptance/openstack/messaging/v2/messaging.go [hunk 3]
```diff
@@ -52,3 +52,16 @@ func GetQueue(t *testing.T, client *gophercloud.ServiceClient, queueName string)
 	}
 	return queue, nil
 }
+
+func CreateShare(t *testing.T, client *gophercloud.ServiceClient, queueName string, clientID string) (queues.QueueShare, error) {
+	t.Logf("Attempting to create share for queue: %s", queueName)
+
+	shareOpts := queues.ShareOpts{
+		Paths:   []string{"messages"},
+		Methods: []string{"POST"},
```
Comment (jtopjian) [inline acceptance/openstack/messaging/v2/messaging.go]: Can these be updated to use the new types?

File: openstack/messaging/v2/queues/requests.go [hunk 4]
```diff
@@ -196,3 +196,55 @@ func GetStats(client *gophercloud.ServiceClient, queueName string) (r StatResult
 		OkCodes: []int{200}})
 	return
 }
+
+type SharePath string
+
+const (
+	ShareMessages      SharePath = "messages"
+	ShareClaims        SharePath = "claims"
+	ShareSubscriptions SharePath = "subscriptions"
```
Comment (jtopjian) [inline openstack/messaging/v2/queues/requests.go]: Slight nit about the naming: I think `PathMessages`, etc would work better:

```go
[]queues.SharePath{queues.PathMessages}
```

**Comment by problemv** [comment]
@jtopjian Sorry about that accidentally let the acceptance test slip thought the cracks on that one. Changes implemented.

**Pull request merged**